tests/test_filer_ftp_pytest.py

```diff
@@ -0,0 +1,221 @@
+            port=login_dict['port']
+            )
+        )
+    mocker.patch(
+        'ftplib.FTP.login',
+        side_effect=conn.login(login_dict['user'], login_dict['passwd'])
+        )
+    mocker.patch('ftplib.FTP.pwd', side_effect=conn.pwd)
+    mocker.patch('ftplib.FTP.cwd', side_effect=conn.cwd)
+    mocker.patch('ftplib.FTP.mkd', side_effect=conn.mkd)
+    mock_storbinary = mocker.patch('ftplib.FTP.storbinary')
+
+    ftp_obj = FTPTransput(
+        "dir1",
+        "ftp://" + login_dict['host'] + "/dir1",
+        Type.Directory,
+        ftp_conn=conn
+    )
+
+    ftp_obj.upload_dir()
+
+    # We use mock.ANY since the 2nd argument of the 'ftplib.FTP.storbinary' is
+    # a file object and we can't have the same between the original and the
+    # mock calls
+    assert sorted(mock_storbinary.mock_calls) == sorted([
+        mock.call('STOR /' + '/dir1/file1', mock.ANY),
+        mock.call('STOR /' + '/dir1/dir2/file2', mock.ANY),
+        mock.call('STOR /' + '/dir1/dir2/file4.txt', mock.ANY)
+    ])
+
+
+def test_ftp_download_dir(mocker, tmpdir, tmp_path, ftpserver):
+    """ Check whether the download of a directory through FTP completes
+        successfully. """
+
+    # Temporary nested directories with files
+    file1 = tmpdir.mkdir("dir1").join("file1")
+    file1.write("this is random")
+    file2 = tmpdir.mkdir("dir1/dir2").join("file2")
+    file2.write('not really')
+    file3 = tmpdir.join('dir1/dir2/file3')
+    file3.write('took me a while')
+
+    # Temporary folder for download
+    tmpdir.mkdir('downloads')
+
+    # Populate the server with the above files to later download
+    ftpserver.put_files({
+        'src': str(tmp_path) + '/dir1/file1',
+        'dest': 'remote1/file1'
+        })
+    ftpserver.put_files({
+        'src': str(tmp_path) + '/dir1/dir2/file2',
+        'dest': 'remote1/remote2/file2'
+        })
+    ftpserver.put_files({
+        'src': str(tmp_path) +  '/dir1/dir2/file3',
+        'dest': 'remote1/remote2/file3'
+        })
+
+    login_dict = ftpserver.get_login_data()
+
+    conn = ftplib.FTP()
+    conn.connect(host=login_dict['host'], port=login_dict['port'])
+    conn.login(login_dict['user'], login_dict['passwd'])
+
+    mock_retrbinary = mocker.patch(
+        'ftplib.FTP.retrbinary',
+        side_effect=conn.retrbinary
+        )
+
+    ftp_obj = FTPTransput(
+        str(tmp_path) + "downloads",
+        "ftp://" + login_dict['host'],
+        Type.Directory,
+        ftp_conn=conn
+        ) 
+
+    ftp_obj.download_dir()
+
+    # We use mock.ANY since the 2nd argument of the 'ftplib.FTP.storbinary' is
+    # a file object and we can't have the same between the original and the
+    # mock calls
+    assert sorted(mock_retrbinary.mock_calls) == sorted([
+        mock.call('RETR ' + '/remote1/file1', mock.ANY),
+        mock.call('RETR ' + '/remote1/remote2/file2', mock.ANY),
+        mock.call('RETR ' + '/remote1/remote2/file3', mock.ANY)      
+    ])
+

```

Comment:
Can we do asserts here for the fact that the files were created in the (temp) file system?

Reply:
Done.